### PR TITLE
Add disable-user command

### LINKED ISSUE #181: Remove accounts for those users who've been removed from the developers list
We've recently removed a few people from the [developers](https://github.com/opensafely-core/backend-server/blob/main/developers) list as they no longer require that level of access to the secure environments. There are some manual steps required for this to take effect on on TPP and EMIS.

Specific users to verify removal:
- [ ] Will J Hulme
- [ ] Seb Bacon
- [ ] ...?

### PR DESCRIPTION
**Pull request by bloodearnest**
This removes all permissions from the user, and the ability to ssh/login
as user. Requires that the user has been removed from all user files
first.

Fixes #181

### CODE REVIEW

**Comment by madwort** [comment]
hmm, so the workflow is:
* remove user from `developers`/`reviewers` etc files
* manually run `remove` command?

what would you say to the workflow as:
* add user to `/removed-users` file which takes precedence over any other file
* also remove user from `developers`/`reviewers` etc files as a tidy-up, but actually doesn't have any effect
* update users command verifies & removes any users listed in `/removed-users` 

Would need a little bit more work (so that the removal operation completes successfully if the user has already been removed), but ultimately more automated, more auditable, less room for forgetting to do something. 

Also, could we add a CI test for the removal? we already test adding a user, right? we could chain the two together?

**Comment by bloodearnest** [comment]
> hmm, so the workflow is:
> 
> * remove user from `developers`/`reviewers` etc files
> * manually run `remove` command?
> 
> what would you say to the workflow as:
> 
> * add user to `/removed-users` file which takes precedence over any other file
> * also remove user from `developers`/`reviewers` etc files as a tidy-up, but actually doesn't have any effect
> * update users command verifies & removes any users listed in `/removed-users`

That's actually quite complex.

1. We may want to remove users from a specific backend, not from all backends. So we'd need multiple removed-users files, one per backend
2. Cleaning removing a user requires adding to one file *and* removing from another, which feels awkward.

I went this simple route after trying to do a more complex approach. Lucy pointed out we don't do it very often, and simple manual tooling + docs is enough to comply with have a removal process, without the complexity tracking state.

> Would need a little bit more work (so that the removal operation completes successfully if the user has already been removed), but ultimately more automated, more auditable, less room for forgetting to do something.

I don't think the audit is much different - there's still a commit removing that user.
 
> Also, could we add a CI test for the removal? we already test adding a user, right? we could chain the two together?

We can, but it's very awkward.  We'd have to modify one of the user files.  On linux, those are bind mounted, so you're modifying the checkout.

I tested in manually in one of the test VMs.  This is an uncommon enough event that I'm not sure the complexity of testing is worth it.

**Comment by madwort** [comment]
> > hmm, so the workflow is:
> > 
> > * remove user from `developers`/`reviewers` etc files
> > * manually run `remove` command?
> > 
> > what would you say to the workflow as:
> > 
> > * add user to `/removed-users` file which takes precedence over any other file
> > * also remove user from `developers`/`reviewers` etc files as a tidy-up, but actually doesn't have any effect
> > * update users command verifies & removes any users listed in `/removed-users`
> 
> That's actually quite complex.
> 
> 1. We may want to remove users from a specific backend, not from all backends. So we'd need multiple removed-users files, one per backend

Ah, ok, I hadn't realised that was a possibility. I was thinking mostly about the case when someone leaves the organisation completely, and/or is a Bennett developer (which is set centrally, not per-backend).

> 2. Cleaning removing a user requires adding to one file *and* removing from another, which feels awkward.

It is a bit double-handling, but I'm interested in having this process run in an automated fashion, and I think it's the easiest way to do that. (The alternative that immediately springs to mind is to have some way to enumerate the users on the server & remove any unexpected users - which feels **way** more complex & error-prone). 

> I went this simple route after trying to do a more complex approach. Lucy pointed out we don't do it very often, and simple manual tooling + docs is enough to comply with have a removal process, without the complexity tracking state.

It's true that we haven't needed to do this very often, but my guess is that we'll need to do it more often in future, as the project as a whole matures into a steady state with some natural churn of people. 

> > Would need a little bit more work (so that the removal operation completes successfully if the user has already been removed), but ultimately more automated, more auditable, less room for forgetting to do something.
> 
> I don't think the audit is much different - there's still a commit removing that user.

Yep, you're right that the commit means that if we do an audit we can know whether or not that user is supposed to have access & check that against ground truth. But whether the command was run correctly on the correct server is just up to the sysadmin & some kind of leaver's process? or other undefined process if someone is changing roles & switching backend? If it was automated we can start thinking about things like running the update_users command on a cron (daily/weekly), and/or an ad-hoc historical audit trail of what access people had at what time based on logs of when the update_users command was run. 

> > Also, could we add a CI test for the removal? we already test adding a user, right? we could chain the two together?
> 
> We can, but it's very awkward.  We'd have to modify one of the user files.  On linux, those are bind mounted, so you're modifying the checkout. 

for developers the update users script [takes a param for the file](https://github.com/opensafely-core/backend-server/blob/main/tests/install.sh#L16), so we wouldn't need to modify a file on disk, we could just pass in a different filename out of the test fixtures, like:

```
developers=test/developers-plus-extra ./scripts/update-users.sh test
# verify extra user exists
developers=test/developers  ./scripts/update-users.sh test
# verify extra user does not exist
```

Obviously we'd have to do some refactoring to be able to do that for the other files like `mybackend/researchers` etc.

> I tested in manually in one of the test VMs.  This is an uncommon enough event that I'm not sure the complexity of testing is worth it.

Yeah, I feel like it might become more common in future, but it has been very uncommon up to now.

**Comment by bloodearnest** [comment]
Some additional context for this PR, when I discussed with @lucyb - tracking the state is hard and complex, so she suggested doing a manual command for now (as we have people waiting to remove) so unblock it, and we can look at automating it more in future if needed.

File: tests/backends/test.sh [hunk 1]
```diff
@@ -22,13 +22,38 @@ just manage
 # run again to check for idempotency
 just manage
 
+# test disable user
+#
+# create a user using a throwaawy developers file
```
Comment (madwort) [inline tests/backends/test.sh:27]:
```suggestion
# create a user using a throwaway developers file
```

**Pull request merged**